### PR TITLE
etcdserver: not reuse connections for peer transport

### LINKED ISSUE #3723: fail to get version of peers in master branch
```
2015-10-20 10:12:04.256264 I | etcdserver: published {Name:infra2 ClientURLs:[http://127.0.0.1:22379]} to cluster ef37ad9dc622a7c4
2015-10-20 10:12:09.258888 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:12380) of member 8211f1d0f64f3269 (Get http://127.0.0.1:12380/version: read tcp 127.0.0.1:49441->127.0.0.1:12380: read: connection reset by peer)
2015-10-20 10:12:09.258919 W | etcdserver: cannot get the version of member 8211f1d0f64f3269 (Get http://127.0.0.1:12380/version: read tcp 127.0.0.1:49441->127.0.0.1:12380: read: connection reset by peer)
2015-10-20 10:12:59.276903 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:12380) of member 8211f1d0f64f3269 (Get http://127.0.0.1:12380/version: net/http: transport closed before response was received)
2015-10-20 10:12:59.276932 W | etcdserver: cannot get the version of member 8211f1d0f64f3269 (Get http://127.0.0.1:12380/version: net/http: transport closed before response was received)
2015-10-20 10:13:19.283007 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:12380) of member 8211f1d0f64f3269 (Get http://127.0.0.1:12380/version: EOF)
2015-10-20 10:13:19.283037 W | etcdserver: cannot get the version of member 8211f1d0f64f3269 (Get http://127.0.0.1:12380/version: EOF)
```

Start 3-member etcd cluster built from the master, and it cannot get version of peers. It may be related to the change that we generate transporter inside etcdserver, which is not set to be the same as before.

### PR DESCRIPTION
**Pull request by yichengq**
etcd uses timeout listener, and times out the accepted connections
if there is no activity. If we use idle connections in
transport, the connection may have timed out, and return error when
doing any request. So we do not reuse connections.

This fixes the problem that etcd fail to get version of peers.

fixes #3723

### CODE REVIEW

**Comment by xiang90** [comment]
Why does it work previously? The peer listener has always been time out one, no?

**Comment by yichengq** [comment]
> Why does it work previously?

The peer transporter was a timeout transport, which sets maxIdleConnections as -1 for itself.

> The peer listener has always been time out one, no? 

yes

**Comment by xiang90** [comment]
then why do not create a timeout transport instead? also can you please fix all the transport that works with peer url?

**Comment by yichengq** [comment]
> then why do not create a timeout transport instead?

it doesn't need read timeout and write timeout in connection because it is not a stream. I could create a timeout transport with no read/write timeout though.

**Comment by xiang90** [comment]
@yichengq 

> it doesn't need read timeout and write timeout in connection because it is not a stream.

Why only stream can have read/write timeout?

**Comment by yichengq** [comment]
> Why only stream can have read/write timeout?

The read/write timeout has the assumption that following reads/writes will happen soon.
And it is general to use request timeout or ResponseHeaderTimeout instead of read/write timeout for HTTP request IMO.

**Comment by xiang90** [comment]
> The read/write timeout has the assumption that following reads/writes will happen soon.
And it is general to use request timeout or ResponseHeaderTimeout instead of read/write timeout for HTTP request IMO.

I do not think you are solving the root problem. The root reason is that the underlying listener for peer http is a timeout listener. It expects the transport to send/recv fast and never reuse the connection. We built the other pair of transport for it already. We really should just use it instead of creating something new with some new reasons. 

Please use the correct pairs of transport/listener. Thanks.

**Comment by yichengq** [comment]
> We built the other pair of transport for it already.

ok. They are used as a pair.

So I changes to use timeoutTransport, and fixes other transport that has the same problem. PTAL.

**Comment by yichengq** [comment]
Per xiang's suggestion, we want to avoid the same bug in the future.

New commit moves the code to create listener and transport for peer communication
at the same place, and use explicit functions to build them. This prevents
possible development errors in the future.

File: rafthttp/transport.go [hunk 1]
```diff
@@ -97,6 +99,28 @@ type Transporter interface {
 	Stop()
 }
 
+// NewPeerListener returns a listener for raft message transfer between peers.
+// It uses timeout listener to identify broken streams promptly.
+func NewPeerListener(u url.URL, tlsInfo transport.TLSInfo) (net.Listener, error) {
```
Comment (xiang90) [inline rafthttp/transport.go]: Can we call this NewRaftListener?

File: rafthttp/transport.go [hunk 2]
```diff
@@ -97,6 +99,28 @@ type Transporter interface {
...
+	return transport.NewTimeoutListener(u.Host, u.Scheme, tlsInfo, ConnReadTimeout, ConnWriteTimeout)
+}
+
+// NewPeerTransport returns a HTTP transport used to send requests to rafthttp listener
+// of remote peers.
+// It uses timeout transport to pair with remote timeout listeners.
+func NewPeerTransport(tlsInfo transport.TLSInfo, dialTimeout time.Duration) (*http.Transport, error) {
```
Comment (xiang90) [inline rafthttp/transport.go]: Can we call this NewRaftTransport?

File: rafthttp/transport.go (hunk 1 above)
Comment (yichengq) [inline rafthttp/transport.go]: sure.

File: rafthttp/transport.go [hunk 3]
```diff
@@ -97,6 +99,28 @@ type Transporter interface {
...
+	return transport.NewTimeoutTransport(tlsInfo, dialTimeout, 0, 0)
+}
+
+// newStreamPeerTransport returns a HTTP transport used to send stream requests
+// to rafthttp listener of remote peers.
+// Read/write timeout is set for stream transport to promptly
+// find out broken status, which minimizes the number of messages
+// sent on broken connection.
+func newStreamPeerTransport(tlsInfo transport.TLSInfo, dialTimeout time.Duration) (*http.Transport, error) {
```
Comment (xiang90) [inline rafthttp/transport.go]: Can we call this newRaftStreamTransport?

**Comment by yichengq** [comment]
I changed the function name to `NewHTTPTransport` because it returns a HTTP transport to do request targeting remote rafthttp listener.

We thought about the idea that makes the new function named `NewTransport`, and changes the naming of existing Transport struct. But that is a super time-consuming one, so I would like to give the new function a different name.

PTAL.

**Comment by yichengq** [comment]
@xiang90 points out that `NewHTTPTransport` still may bring confusing considering rafthttp has `Transport` struct already. I am targetting to return RoundTripper instead.

**Comment by yichengq** [comment]
@xiang90 Changed it to `NewRoundTripper` and this eliminates confusion of duplicate Transport word. PTAL.

**Comment by yichengq** [comment]
Based on discussion OOL, moved new functions into util.go and add comments about 0 read/write timeout.

**Comment by xiang90** [comment]
LGTM

**Pull request merged**